### ISSUE TITLE
[Generator Crash Bug] Type 'List<String?>' is not a subtype of type 'List<String>'

### ISSUE THREAD

**Issue by mozochek** [Milad-Akarie/injectable#205]
## Summary
This bug occurs only after updating 'build_runner' from ^1.12.2 to highest vesions. Versions of 'injectable' and 'injectable_generator' is not important. With 'build_runner' ^1.12.2 and lowest versions everything is ok.

## Steps to reproduce
1. Add 'build_runner' ^2.0.0 and higher, 'injectbale' and 'injectable_generator' with correct versions.
2. Inject someting.
3. Run ```flutter pub run build_runner build```

<details>
<summary>minimal code sample (actually it doesn't matter which code to run, but anyway)</summary>

```dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:injectable/injectable.dart';

@module
abstract class FirebaseFirestoreInjectableModule {
  @Singleton(
    env: <String>[
      Environment.test,
      Environment.dev,
      Environment.prod,
    ],
  )
  FirebaseFirestore get firebaseFirestore => FirebaseFirestore.instance;
}

@module
abstract class FirebaseAuthInjectableModule {
  @Singleton(
    env: <String>[
      Environment.test,
      Environment.dev,
      Environment.prod,
    ],
  )
  FirebaseAuth get firebaseAuth => FirebaseAuth.instance;
}

```
</details>

<details>
<summary>flutter doctor --verbose</summary>

```
[√] Flutter (Channel stable, 2.0.5, on Microsoft Windows [Version 10.0.19042.928], locale ru-RU)
    • Flutter version 2.0.5 at C:\Important\Programs\flutter
    • Framework revision adc687823a (12 days ago), 2021-04-16 09:40:20 -0700
    • Engine revision b09f014e96
    • Dart version 2.12.3

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Kirill\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: C:\Important\Programs\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[X] Chrome - develop for the web (Cannot find Chrome executable at .\Google\Chrome\Application\chrome.exe)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.9.4)
    • Visual Studio at C:\Important\Programs\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.9.31205.134
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Important\Programs\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at C:\Important\Programs\IntelliJ IDEA Community Edition
    • Flutter plugin version 55.1.5
    • Dart plugin version 211.7233

[√] Connected device (2 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19042.928]
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 90.0.818.49

! Doctor found issues in 1 category.
```

</details>

## Where bug occur
To get more information about what happend I run ```flutter pub run build_runner build --verbose --delete-conflicting-outputs``` and get this log:

<details>
<summary>log</summary>

```
flutter pub run build_runner build --verbose --delete-conflicting-outputs
[ +169 ms] executing: [C:\Important\Programs\flutter/] git -c log.showSignature=false log -n 1 --pretty=format:%H
[ +122 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[   +1 ms] adc687823a831bbebe28bdccfac1a628ca621513
[   +1 ms] executing: [C:\Important\Programs\flutter/] git tag --points-at adc687823a831bbebe28bdccfac1a628ca621513
[  +93 ms] Exit code 0 from: git tag --points-at adc687823a831bbebe28bdccfac1a628ca621513
[        ] 2.0.5
[  +81 ms] executing: [C:\Important\Programs\flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +69 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/stable
[        ] executing: [C:\Important\Programs\flutter/] git ls-remote --get-url origin
[  +60 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[ +134 ms] executing: [C:\Important\Programs\flutter/] git rev-parse --abbrev-ref HEAD
[  +62 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] stable
[ +193 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +6 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +119 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'GradleWrapper' is not required, skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +1 ms] Artifact Instance of 'FlutterSdk' is not required, skipping update.
[        ] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'PubDependencies' is not required, skipping update.
[  +77 ms] executing: C:\Important\Programs\flutter\bin\cache\dart-sdk\bin\pub.bat run build_runner build --verbose --delete-conflicting-outputs
[INFO] Entrypoint:Generating build script...
[INFO] Entrypoint:Generating build script completed, took 926ms

[INFO] BuildDefinition:Initializing inputs
[INFO] BuildDefinition:Reading cached asset graph...
[INFO] BuildDefinition:Reading cached asset graph completed, took 223ms

[INFO] BuildDefinition:Checking for updates since last build...
[INFO] BuildDefinition:Checking for updates since last build completed, took 1.9s

[INFO] Build:Running build...
[INFO] Build:Running build completed, took 27ms

[INFO] Build:Caching finalized dependency graph...
[INFO] Build:Caching finalized dependency graph completed, took 95ms

[SEVERE] injectable_generator:injectable_builder on lib/src/internal/di/firebase_injectable_modules.dart (cached):

type 'List<String?>' is not a subtype of type 'List<String>'
package:injectable_generator/resolvers/dependency_resolver.dart 152:5    DependencyResolver._resolveActualType
package:injectable_generator/resolvers/dependency_resolver.dart 91:12    DependencyResolver.resolveModuleMember
package:injectable_generator/generators/injectable_generator.dart 51:15  InjectableGenerator.generate
package:source_gen/src/builder.dart 327:23                               _generate

[SEVERE] Build:
Failed after 203ms
[+33419 ms] "flutter run" took 33 715ms.
[  +12 ms] pub finished with exit code 1
[   +3 ms] 
           #0      throwToolExit (package:flutter_tools/src/base/common.dart:10:3)
           #1      _DefaultPub.interactively (package:flutter_tools/src/dart/pub.dart:364:7)
           <asynchronous suspension>
           #2      PackagesForwardCommand.runCommand (package:flutter_tools/src/commands/packages.dart:238:5)
           <asynchronous suspension>
           #3      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:1157:12)
           <asynchronous suspension>
           #4      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1009:27)
           <asynchronous suspension>
           #5      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
           <asynchronous suspension>
           #6      AppContext.run (package:flutter_tools/src/base/context.dart:149:12)
           <asynchronous suspension>
           #7      CommandRunner.runCommand (package:args/command_runner.dart:197:13)
           <asynchronous suspension>
           #8      FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:278:9)
           <asynchronous suspension>
           #9      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
           <asynchronous suspension>
           #10     AppContext.run (package:flutter_tools/src/base/context.dart:149:12)
           <asynchronous suspension>
           #11     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:234:5)
           <asynchronous suspension>
           #12     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:64:9)
           <asynchronous suspension>
           #13     run.<anonymous closure> (package:flutter_tools/runner.dart:62:12)
           <asynchronous suspension>
           #14     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
           <asynchronous suspension>
           #15     AppContext.run (package:flutter_tools/src/base/context.dart:149:12)
           <asynchronous suspension>
           #16     runInContext (package:flutter_tools/src/context_runner.dart:73:10)
           <asynchronous suspension>
           #17     main (package:flutter_tools/executable.dart:90:3)
           <asynchronous suspension>
           
           
[ +221 ms] ensureAnalyticsSent: 213ms
[   +6 ms] Running shutdown hooks
[   +1 ms] Shutdown hooks complete
[   +1 ms] exiting with code 1

```

</details>
From all log lines we need only few of them:

```
type 'List<String?>' is not a subtype of type 'List<String>'
package:injectable_generator/resolvers/dependency_resolver.dart 152:5    DependencyResolver._resolveActualType
package:injectable_generator/resolvers/dependency_resolver.dart 91:12    DependencyResolver.resolveModuleMember
package:injectable_generator/generators/injectable_generator.dart 51:15  InjectableGenerator.generate
package:source_gen/src/builder.dart 327:23                               _generate
```

We can see that error starts in `dependency_resolver.dart` in 152 line. I check the source code to see what actually happend.

```dart
 35  List<String> _environments = [];
...
107  var inlineEnv;
...
132  inlineEnv = injectable
133    .peek('env')
134    ?.listValue
135    .map((e) => e.toStringValue())
136     .toList();
...
152  _environments = inlineEnv ??
153    _envChecker
154      .annotationsOf(annotatedElement)
155   .map<String>(
156      (e) => e.getField('name')!.toStringValue()!,
157    )
158    .toList() ??
159    const [];
```
In code above we assign to '_environments' value of 'inlineEnv'. The problem is that type of '_environments' is `List<String>`, but the type of 'inlineEnv' is `List<String?>`, because in 135 line we call `e.toStringValue()` which return `String?` type.

**Comment by proninyaroslav** [comment]
Same here

**Comment by proninyaroslav** [comment]
As a workaround:
```diff
diff --git a/injectable_generator/lib/resolvers/dependency_resolver.dart b/injectable_generator/lib/resolvers/dependency_resolver.dart
index c85d459..52667f6 100644
--- a/injectable_generator/lib/resolvers/dependency_resolver.dart
+++ b/injectable_generator/lib/resolvers/dependency_resolver.dart
@@ -132,7 +132,7 @@ class DependencyResolver {
       inlineEnv = injectable
           .peek('env')
           ?.listValue
-          .map((e) => e.toStringValue())
+          .map((e) => e.toStringValue()!)
           .toList();
     }
     if (abstractType != null) {
```

**Comment by proninyaroslav** [comment]
@Milad-Akarie 
When will this be fixed?

**Comment by theweiweiway** [comment]
Just started getting this too~

**Comment by theweiweiway** [comment]
hey @proninyaroslav , I get this error when i make that change:

```dart
Null check operator used on a null value
package:build_runner/src/generate/watch_impl.dart 189:40  WatchImpl.reader
package:build_runner/src/server/server.dart 50:55         createServeHandler.<fn>
package:build_runner/src/server/server.dart 49:20         createServeHandler.<fn>
```

Do you also get this error?

**Comment by proninyaroslav** [comment]
@theweiweiway 
No, I haven't encountered this problem. Maybe this is different problem.

**Comment by SSebigo** [comment]
@theweiweiway Do you have an idea for when will the version with your fix be release?

**Comment by theweiweiway** [comment]
I closed my PR , since I tried it and it didn't work :S

**Comment by proninyaroslav** [comment]
@theweiweiway 
This works fine for me :D. Probably `injectable-generator` has several bugs that appears under certain conditions.

**Comment by rmargolinross** [comment]
I have run into this issue as well. I need a work around or for something to get fixed. This is blocking any builds.

**Comment by mozochek** [comment]
https://github.com/Milad-Akarie/injectable/pull/211

**Comment by Milad-Akarie** [comment]
Fixed in version 1.4.1
sorry for the delay.
